RedHatInsights/insights-rbac
# Disallow changing target org

## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-19783

## Description of Intent of Change(s)
- API calls to update CARs are removing the target_org. This change will fix the restriction to prevent this from happening.

## Local Testing
A new unit test now tests that we can't change the target_org of a CAR.
To test locally:
- Create a CAR:
```
curl -X 'POST' \
  'https://editor.swagger.io/api/rbac/v1/cross-account-requests/' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "target_account": "12345",
  "target_org": "12345",
  "start_date": "01/01/2021",
  "end_date": "01/01/2021",
  "roles": [
    "Role Name"
  ]
}'
```
- Test updating the CARs target_org
```
curl -X 'PUT'   'http://localhost:8000/api/rbac/v1/cross-account-requests/<uuid>/'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "start_date": "07/11/2022",
  "end_date": "07/21/2022",
  "roles": [
    "Catalog User Local Test"
  ],
  "target_org": "132123"
}'
```
- You should receive an error:
`{"errors":[{"detail":"Target org must stay the same.","source":"cross-account-update","status":"400"}]}`
- Test updating CARs target_org with NULL
```
curl -X 'PUT'   'http://localhost:8000/api/rbac/v1/cross-account-requests/<uuid>/'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "start_date": "07/11/2022",
  "end_date": "07/21/2022",
  "roles": [
    "Catalog User Local Test"
  ],
  "target_org": ""
}'
```
- You should receive an error:
`{"errors":[{"detail":"This field may not be blank.","source":"target_org","status":"400"}]}`
- Test PATCHing a CAR with a new target_org
```
curl -X 'PATCH'   'http://localhost:8000/api/rbac/v1/cross-account-requests/<uuid>/'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "start_date": "07/11/2022",
  "end_date": "07/21/2022",
  "roles": [
    "Catalog User Local Test"
  ],
  "status": "cancelled",
  "target_org": "21124"
}'
```
- You should receive an error:
`{"errors":[{"detail":"Field 'target_org' is not supported. Please use one or more of: ['start_date', 'end_date', 'roles', 'status']","source":"cross-account partial update","status":"400"}]}`

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [X] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  -  if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
